# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report,ConfusionMatrixDisplay,confusion_matrix,balanced_accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

## Split the Data into Training and Testing Sets

In [3]:
data.isnull().sum()

word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 

In [5]:
data.describe()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
count     4601.000000        4601.000000    4601.000000   4601.000000   
mean         0.104553           0.213015       0.280656      0.065425   
std          0.305358           1.290575       0.504143      1.395151   
min          0.000000           0.000000       0.000000      0.000000   
25%          0.000000           0.000000       0.000000      0.000000   
50%          0.000000           0.000000       0.000000      0.000000   
75%          0.000000           0.000000       0.420000      0.000000   
max          4.540000          14.280000       5.100000     42.810000   

       word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
count    4601.000000     4601.000000       4601.000000         4601.000000   
mean        0.312223        0.095901          0.114208            0.105295   
std         0.672513        0.273824          0.391441            0.401071   
min         0.000000        0.000000          0.000000            0.000000   
25%         0.000000        0.000000          0.000000            0.000000   
50%         0.000000        0.000000          0.000000            0.000000   
75%         0.380000        0.000000          0.000000            0.000000   
max        10.000000        5.880000          7.270000           11.110000   

       word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
count      4601.000000     4601.000000  ...  4601.000000  4601.000000   
mean          0.090067        0.239413  ...     0.038575     0.139030   
std           0.278616        0.644755  ...     0.243471     0.270355   
min           0.000000        0.000000  ...     0.000000     0.000000   
25%           0.000000        0.000000  ...     0.000000     0.000000   
50%           0.000000        0.000000  ...     0.000000     0.065000   
75%           0.000000        0.160000  ...     0.000000     0.188000   
max           5.260000       18.180000  ...     4.385000     9.752000   

       char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
count  4601.000000  4601.000000  4601.000000  4601.000000   
mean      0.016976     0.269071     0.075811     0.044238   
std       0.109394     0.815672     0.245882     0.429342   
min       0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.315000     0.052000     0.000000   
max       4.081000    32.478000     6.003000    19.829000   

       capital_run_length_average  capital_run_length_longest  \
count                 4601.000000                 4601.000000   
mean                     5.191515                   52.172789   
std                     31.729449                  194.891310   
min                      1.000000                    1.000000   
25%                      1.588000                    6.000000   
50%                      2.276000                   15.000000   
75%                      3.706000                   43.000000   
max                   1102.500000                 9989.000000   

       capital_run_length_total         spam  
count               4601.000000  4601.000000  
mean                 283.289285     0.394045  
std                  606.347851     0.488698  
min                    1.000000     0.000000  
25%                   35.000000     0.000000  
50%                   95.000000     0.000000  
75%                  266.000000     1.000000  
max                15841.000000     1.000000  

[8 rows x 58 columns]

In [6]:
# Create the labels set `y` and features DataFrame `X`
X=data.drop(columns=["spam"],axis=1)
X

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0               0.00               0.64           0.64           0.0   
1               0.21               0.28           0.50           0.0   
2               0.06               0.00           0.71           0.0   
3               0.00               0.00           0.00           0.0   
4               0.00               0.00           0.00           0.0   
...              ...                ...            ...           ...   
4596            0.31               0.00           0.62           0.0   
4597            0.00               0.00           0.00           0.0   
4598            0.30               0.00           0.30           0.0   
4599            0.96               0.00           0.00           0.0   
4600            0.00               0.00           0.65           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0              0.32            0.00              0.00                0.00   
1              0.14            0.28              0.21                0.07   
2              1.23            0.19              0.19                0.12   
3              0.63            0.00              0.31                0.63   
4              0.63            0.00              0.31                0.63   
...             ...             ...               ...                 ...   
4596           0.00            0.31              0.00                0.00   
4597           0.00            0.00              0.00                0.00   
4598           0.00            0.00              0.00                0.00   
4599           0.32            0.00              0.00                0.00   
4600           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0                0.00            0.00  ...                   0.0        0.000   
1                0.00            0.94  ...                   0.0        0.000   
2                0.64            0.25  ...                   0.0        0.010   
3                0.31            0.63  ...                   0.0        0.000   
4                0.31            0.63  ...                   0.0        0.000   
...               ...             ...  ...                   ...          ...   
4596             0.00            0.00  ...                   0.0        0.000   
4597             0.00            0.00  ...                   0.0        0.000   
4598             0.00            0.00  ...                   0.0        0.102   
4599             0.00            0.00  ...                   0.0        0.000   
4600             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0           0.000          0.0        0.778        0.000        0.000   
1           0.132          0.0        0.372        0.180        0.048   
2           0.143          0.0        0.276        0.184        0.010   
3           0.137          0.0        0.137        0.000        0.000   
4           0.135          0.0        0.135        0.000        0.000   
...           ...          ...          ...          ...          ...   
4596        0.232          0.0        0.000        0.000        0.000   
4597        0.000          0.0        0.353        0.000        0.000   
4598        0.718          0.0        0.000        0.000        0.000   
4599        0.057          0.0        0.000        0.000        0.000   
4600        0.000          0.0        0.125        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
0                          3.756                          61   
1                          5.114                         101   
2                          9.821                         485   
3                          3.537                          40   
4                          3.537              

In [7]:
y=data["spam"]
y

0       1
1       1
2       1
3       1
4       1
       ..
4596    0
4597    0
4598    0
4599    0
4600    0
Name: spam, Length: 4601, dtype: int64

In [8]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
data["spam"].value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [9]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test=train_test_split(X,y,random_state=42)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [10]:
# Create the StandardScaler instance
scaler=StandardScaler()

In [11]:
# Fit the Standard Scaler with the training data
# Scale the training data
X_train_scaled=scaler.fit_transform(X_train)
X_train_scaled

array([[-0.33933373, -0.16214839, -0.55144324, ..., -0.11649103,
        -0.21175305, -0.41909144],
       [ 1.8777157 , -0.16214839, -0.55144324, ..., -0.10814021,
        -0.18841316, -0.35960578],
       [-0.33933373, -0.16214839,  2.2325961 , ...,  0.21993678,
         0.29705664, -0.10941609],
       ...,
       [-0.33933373, -0.16214839, -0.55144324, ..., -0.08642808,
        -0.20241709, -0.40334524],
       [-0.33933373,  0.02143809, -0.55144324, ..., -0.10902256,
        -0.18841316, -0.18639753],
       [-0.00843083, -0.16214839,  0.25239911, ..., -0.09960031,
        -0.1137255 ,  0.10403246]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [12]:
# Train a Logistic Regression model and print the model score

Model=LogisticRegression(random_state=1)
Model.fit(X_train_scaled,y_train)

LogisticRegression(random_state=1)

In [13]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions=Model.predict(scaler.transform(X_test))
# Review the predictions
testing_predictions

array([0, 0, 0, ..., 1, 0, 0], dtype=int64)

In [14]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
print(accuracy_score(y_test,testing_predictions))

0.9226759339704604


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [15]:
# Train a Random Forest Classifier model and print the model score
rd=RandomForestClassifier(n_estimators=100,random_state=1)
rd.fit(X_train_scaled,y_train)


RandomForestClassifier(random_state=1)

In [16]:
# Make and save testing predictions with the saved  using the test data
rd_testing_predictions=rd.predict(scaler.transform(X_test))
# Review the predictions
rd_testing_predictions



array([0, 0, 0, ..., 1, 0, 0], dtype=int64)

In [17]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
print(accuracy_score(y_test,rd_testing_predictions))

0.9582971329278888


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

*Replace the text in this markdown cell with your answers to these questions.*
 The Linear Regression Model's accuracy score ended up being 0.92 (92.2% accuracy), whereas the Random Forest Classifier Model's accuracy score was 0.958 (96.0% accuracy). This means the random forest model did end up outperforming the linear regression model by around 4% which is significant. Both models scored above 92% accuracy which is great and means both models predict spam relatively well every time.  The results support that hypothesis. It is clear, however, that both models performed great at classifying the data, and could be awesome models to use outside of class on real life work projects.

In [18]:

import matplotlib.pyplot as plt


In [19]:
print(confusion_matrix(y_test, rd_testing_predictions,labels=[1,0]))

[[438  37]
 [ 11 665]]


In [20]:
print(classification_report(y_test, rd_testing_predictions,labels=[1,0]))

              precision    recall  f1-score   support

           1       0.98      0.92      0.95       475
           0       0.95      0.98      0.97       676

    accuracy                           0.96      1151
   macro avg       0.96      0.95      0.96      1151
weighted avg       0.96      0.96      0.96      1151



In [21]:
print(balanced_accuracy_score(y_test, rd_testing_predictions))

0.9529165369043912


In [22]:
print(confusion_matrix(y_test, testing_predictions,labels=[1,0]))

[[413  62]
 [ 27 649]]


In [23]:
print(classification_report(y_test, testing_predictions,labels=[1,0]))

              precision    recall  f1-score   support

           1       0.94      0.87      0.90       475
           0       0.91      0.96      0.94       676

    accuracy                           0.92      1151
   macro avg       0.93      0.91      0.92      1151
weighted avg       0.92      0.92      0.92      1151



In [24]:
print(balanced_accuracy_score(y_test, testing_predictions))

0.9147664279040797
